[jit] Fix aug assign for non-tensor attributes

driazati opened the pull request:
Stacked PRs
 * #33578 - [jit] Unify augmented assign handling
 * **#32993 - [jit] Fix aug assign for non-tensor attributes**

Instead of erroring out this de-sugars augmented assignments to class
members from `self.a += 1` to `self.a = self.a + 1`.

Fixes #32973

Differential Revision: [D19737636](https://our.internmc.facebook.com/intern/diff/19737636/)

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1871,11 +1870,16 @@ struct to_ir {
-      throw ErrorReport(stmt.lhs())
-          << "left-hand side of augmented assignment to module "
-          << "parameters/buffers can only be tensor types";
+      const auto rhs = NamedValue(stmt.rhs().range(), emitExpr(stmt.rhs())).value(*method.graph());
```
I dont think this correctly handles objects which define `__iadd__` (which is why tensors do an inplace update)

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1871,11 +1886,38 @@ struct to_ir {
+      // https://docs.python.org/3/reference/datamodel.html#object.__iadd__
+      auto magic_method_name = getAugMagicMethod(stmt);
+      const auto rhs = NamedValue(stmt.rhs().range(), emitExpr(stmt.rhs()))
+                           .value(*method.graph());
```
Why do we need to do `NamedValue(stmt.rhs().range(), emitExpr(stmt.rhs())).value(*method.graph())` instead of just `emitExpr` ?

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1854,8 +1870,7 @@ struct to_ir {
   // in place op, and throw error for other unsupported types
   void emitAugAssignmentToSelectVar(const AugAssign& stmt) {
     const auto lhs = Select(stmt.lhs());
-    const auto lhsSugaredVar =
```
I think you need to add handling of Lists `+=`, and `*=` as well.

eellison reviewed (commented):
Looks good! Just needs handling of lists

Commit:
[jit] Add type tags to lists/dicts in pickle

This adds a global call to `torch.jit._pickle.restore_type_tags` for
lists and dicts so that we can preserve their types after serialization.

Commit:
removing changes

Commit:
Instead of erroring out this de-sugars augmented assignments to class
members from `self.a += 1` to `self.a = self.a + 1`.

Fixes #32973

Differential Revision: [D19737636](https://our.internmc.facebook.com/intern/diff/19737636/)

Commit:
Fix list inplace update

Commit:
Merge branch 'master' of github.com:pytorch/pytorch into driazati/fixnest

Commit:
fix flake

eellison commented on test/test_jit.py:
```diff
@@ -5261,6 +5261,48 @@ def foo(z):
+            def __init__(self):
+                self.num = 99
+
+            def __iadd__(self, x):
```
Can you add a test for a class which doesn't define `__iadd__`, but does define `__add__` ? 
```
class A(object):
    def __init__(self, num):
        self.a = num

    def __add__(self, other):
        return A(self.a + other.a)


class B(object):
    def __init__(self, my_a):
        self.my_a = my_a

    def do_plus_equal(self, a_val):
        self.my_a += a_val
        return self.my_a.a

b = B(A(4))
print(b.do_plus_equal(A(5)))
# 9
```

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1871,11 +1881,37 @@ struct to_ir {
+      const auto rhs = emitExpr(stmt.rhs());
+
+      // Insert call to magic method
+      auto sugared_magic_method = makeMagic(
```
I think this also needs to handle classes which define `__add__`.  I added a test case above. The pattern here is:

 Does class define `__iadd__` ?

yes: call it 
no: does class define `__add__` ?

yes: call it
no: error

eellison reviewed (commented):
Looks good, i think there's still one more case to handle. Commented on the code.

Commit:
Fix list

Commit:
Fix iterator invalidation bug

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -1939,6 +1939,26 @@ int listInplaceAdd(Stack& stack) {
 }
 
+template <class T>
+int listMulIntLeftInPlace(Stack& stack) {
```
Why does this need to get implemented this way instead of desugaring 
```
x = [1, 2, 3]
x *= 3
``` 

to:
```
x = [1, 2, 3] * 3
```
?

driazati replied:
Doing it this way avoids copying the elements in the list for 1 of the iterations

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1797,22 +1797,36 @@ struct to_ir {
+  // Get a pair of <in place magic method name, out of place magic method name>
+  // since the out of place method is called if the in place method is not
+  // present
+  std::pair<const char*, const char*> getAugMagicMethod(const AugAssign& stmt) {
```
apparently all string literals are permanently allocated, so this is memory safe. but if someone were to edit this in a way that dynamically creates string literals it no longer would be. since these `const char *` can upcasted in `makeMagic` to `std::string` anyway can you use `std::string` here instead of `const char *`? It is also one less thing to think about when reviewing and maintaining this code.

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1871,11 +1884,48 @@ struct to_ir {
+      auto result = sugared_magic_method
+                        ->call(stmt.range(), method, {lhsValue, rhs}, {}, 0)
+                        ->asValue(stmt.range(), method);
+      // x += y is equivalent to x == x.__iadd__(y), so set the value to the
```
this comment is no longer valid since at this point we could be doing either `__iadd__` or `__add__`

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1871,11 +1884,48 @@ struct to_ir {
+                        ->asValue(stmt.range(), method);
+      // x += y is equivalent to x == x.__iadd__(y), so set the value to the
+      // function's return value
+      lhsSugaredVar->setAttr(
```
it would be better to not add a setAttr node if it's an inplace update, since it's unneeded and adds extra mutation & nodes to the graph.

driazati replied:
You have to `setattr` to match Python's semantics, which use the return value of `__iadd__` to set the new value

eellison replied:
It doesn't break semantics. The identity of the object is the same regardless.

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1871,11 +1884,48 @@ struct to_ir {
           {},
           self);
-
+    } else if (lhsValue->type()->kind() == TypeKind::ClassType) {
```
The comment on the top of this function is now wrong:
```
  // In this case we will only consider the scenario that the module
  // buffer type is a tensor, and we emit the corresponding tensor
  // in place op, and throw error for other unsupported types
```

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1854,8 +1868,7 @@ struct to_ir {
   // in place op, and throw error for other unsupported types
   void emitAugAssignmentToSelectVar(const AugAssign& stmt) {
```
Why is the flow of this function:

```
if (val->isTensor())  {

} else if (val->isClassType()) {

} else {

}
```
Can't we delete the first block ?

eellison commented on torch/csrc/jit/script/ir_emitter.cpp:
```diff
@@ -1871,11 +1884,48 @@ struct to_ir {
+      auto sugared_magic_method = makeMagic(
+          magic_method_name,
+          std::make_shared<BuiltinFunction>(
+              getAugOp(stmt, lhsValue->type()), at::nullopt));
```
`getAugOp` will never be called or used here. There's not really a need to use the magic method sugared values since within this block it will always be a class type. you can just throw if the class doesn't define `in_place_method_name` or `out_of_place_method_name`.

eellison reviewed (approved):
Looks good, this will be a nice change to land. Needs a few changes but they should all be easily addressable.

driazati commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -2554,6 +2574,10 @@ RegisterOperators reg2({
+          aliasAnalysisFromSchema()),                                               \
+      Operator(                                                                     \
+          "aten::mul_(" decl_type "[] l, int n) -> " decl_type "[]",                \
+          listMulIntLeftInPlace<c_type::value_type>,                                \
```
leaving for a follow up since it also requires fixing the template parameters of all the functions that use it

Commit:
Address comments
